_delb/names.py

```diff
@@ -57,27 +58,66 @@ class Namespaces(Mapping):
 
     # https://www.w3.org/TR/xml-names/#xmlReserved
 
-    __slots__ = ("data",)
+    __slots__ = ("__data", "fallback")
 
-    def __init__(self, namespaces: Mapping[Optional[str], str]):
-        self.data: Mapping[Optional[str], str]
+    def __init__(
+        self,
+        namespaces: NamespaceDeclarations,
+        *,
+        fallback: Optional[Namespaces] = None,
+    ):
+        self.__data: NamespaceDeclarations
         if isinstance(namespaces, Namespaces):
-            self.data = namespaces.data
+            self.__data = namespaces.__data
+
+        elif isinstance(namespaces, Mapping):
+            # https://www.w3.org/TR/REC-xml-names seems not to care about redundantly
+            # declared namespaces.
+            self.__data = {}
+            for prefix, namespace in namespaces.items():
+                if prefix in ("xml", "xmlns"):
+                    raise ValueError(
+                        f"One must not override the global prefix `{prefix}`."
+                    )
+
+                if namespace == "":
+                    pass  # empty namespaces are allowed
+                elif namespace in {XML_NAMESPACE, XMLNS_NAMESPACE}:
+                    raise ValueError(
+                        f"The namespace `{namespace}` must not be overridden."
+                    )
+                elif not is_iri_compliant(namespace):
+                    raise ValueError(f"`{namespace}` is not a valid IRI.")
+
+                self.__data[prefix] = namespace
+
         else:
-            assert isinstance(namespaces, Mapping)
-            self.data = namespaces
+            raise TypeError
+
+        self.fallback: NamespaceDeclarations = (
+            {"xml": XML_NAMESPACE, "xmlns": XMLNS_NAMESPACE}
+            if fallback is None
+            else fallback
+        )
 
-    def __getitem__(self, item):
-        return GLOBAL_PREFIXES.get(item) or self.data[item]
+    def __getitem__(self, item) -> str:
+        return self.__data.get(item) or self.fallback[item]
+
+    def __hash__(self):
+        data_hash = hash(frozenset(self.__data.items()))
+        if isinstance(self.fallback, Namespaces):
+            return hash((data_hash, hash(self.fallback)))
+        else:
+            return data_hash
```

Comment:
some may ask: what is this good for?

an immediate use-case will be to cache lookup of prefixes in a node's context for serialization. otherwise one would have to search linearly through possibly multiple mapping values to find a prefix.
generally, i think that immutable objects should generally be hashable, b/c throwing objects into sets is fun (and often the cleanest way to solve a problem).

i should move the calculation to `__init__` though.